openmdao/core/implicitcomponent.py

```diff
@@ -317,6 +356,258 @@ def add_output(self, name, val=1.0, **kwargs):
 
         return metadata
 
+    def add_residual(self, name, shape=(), units=None, desc='', ref=1.0):
+        """
+        Add an residual variable to the component.
+
+        Note that the total size of the residual vector must match the total size of
+        the outputs vector for this component.
+
+        Parameters
+        ----------
+        name : str
+            Name of the residual in this component's namespace.
+        shape : int or tuple
+            Shape of this residual.
+        units : str or None
+            Units in which this residual will be given to the user when requested.
+            Default is None, which means it has no units.
+        desc : str
+            Description of the residual.
+        ref : float or ndarray
+            Scaling parameter. The value in the user-defined units of this residual
```

Comment:
Is this 'ref' the same as 'res_ref' on add_output? What happens if the user declares a 'res_ref' on one of the outputs, and then a different 'ref' here?

Reply:
I added checks for ref/res_ref and units/res_units for residuals and outputs that overlap.